brod/brodrust/src/protocol/codecs.rs

```diff
@@ -0,0 +1,430 @@
+use std::io::{Read, Write};
+use std::default::Default;
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use crate::kafka_error::{Result, ErrorKind};
+
+// Helper macro to safely convert an usize expression into a signed
+// integer.  If the conversion is not possible the macro issues a
+// `CodecError`, otherwise returns the expression
+// in the requested target type.
+macro_rules! try_usize_to_int {
+    // ~ $ttype should actually be a 'ty' ... but rust complains for
+    // some reason :/
+    ($value:expr, $ttype:ident) => {{
+        let maxv = $ttype::max_value();
+        let x: usize = $value;
+        if (x as u64) <= (maxv as u64) {
+            x as $ttype
+        } else {
+            bail!(ErrorKind::CodecError)
+        }
+    }}
+}
+
+pub trait ToByte {
+    fn encode<T: Write>(&self, buffer: &mut T) -> Result<()>;
+}
+
+
+impl<'a, T: ToByte + 'a + ?Sized> ToByte for &'a T {
+    fn encode<W: Write>(&self, buffer: &mut W) -> Result<()> {
+        (*self).encode(buffer)
+    }
+}
+
+impl ToByte for i8 {
+    fn encode<T: Write>(&self, buffer: &mut T) -> Result<()> {
+        buffer.write_i8(*self).or_else(|e| Err(From::from(e)))
```

Comment:
No need to use `or_else(...`, `buffer.write_i8(*self)?` will be better

Reply:
I will remove up error-chain lib and corresponding code in another PR